modules: samples: Add wasm-micro-runtime as module and add samples

wenyongh opened the pull request:
Add wasm-micro-runtime as a Zephyr module and add two samples:
a hello world sample to illustrate how to run WASM application with
wasm-micro-runtime (WAMR) runtime engine, and an app manager
sample to illustrate how to start WAMR app manager in device
side which runs on Zephyr OS and then install/uninstall/query
WASM apps from host side.

Signed-off-by: Wenyong Huang <wenyong.huang@intel.com>

carlescufi commented:
@wenyongh thanks! This seems to address the integration of WASM with Zephyr at the sample level. We should instead try to do something a bit more generic, so that each sample doesn't need to copy over all the boilerplate. What are your thoughts on that?

carlescufi commented:
@xwang98 could you take a look and compare to your approach?

wenyongh commented:
> @wenyongh thanks! This seems to address the integration of WASM with Zephyr at the sample level. We should instead try to do something a bit more generic, so that each sample doesn't need to copy over all the boilerplate. What are your thoughts on that?

Yes, it is really not generic and not convenient to developer, how about adding WAMR as a module to Zephyr? For example, add folder zephyr/modules/wasm-micro-runtime and file Kconfig/CMakeLists.txt/wamr_init.c inside it, and define the related macros to configure WAMR features, e.g., CONFIG_WAMR_INTERP, CONFIG_WAMR_AOT? By default the runtime environment is initialized in wamr_init.c, and developer can simply load and run the wasm application. Thanks.

tejlmand commented on samples/wasm_micro_runtime/hello_world/CMakeLists.txt:
```diff
@@ -0,0 +1,49 @@
+  set (WAMR_BUILD_LIBC_WASI 0)
+endif ()
+
+set (WAMR_ROOT_DIR ${CMAKE_CURRENT_SOURCE_DIR}/../../../../modules/lib/wasm-micro-runtime)
```
If the wasm-micro-runtime is made into a Zephyr library, that is having a: `zephyr/module.yml` file with this content:
```
name: wasm-micro-runtime

build:
  cmake-ext: True
  kconfig-ext: True
```
then we can use `ZEPHYR_WASM_MICRO_RUNTIME_MODULE_DIR` instead.

tejlmand replied:
See also: 
https://github.com/zephyrproject-rtos/loramac-node/blob/zephyr/zephyr/module.yml
https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/loramac-node/CMakeLists.txt

for inspiration.

wenyongh replied:
Done.

tejlmand reviewed (changes requested):
@wenyongh thanks for this contribution. A good start, but needs some additional effort.

I agree with @carlescufi that this should be more generic, and for a start I have placed a few comments for you to take a look at.

wenyongh commented:
> @wenyongh thanks for this contribution. A good start, but needs some additional effort.
> 
> I agree with @carlescufi that this should be more generic, and for a start I have placed a few comments for you to take a look at.

@tejlmand Thanks a lot for your suggestions! We will modify the patch according your comments and try to make it more generic.

wenyongh committed:
Merge pull request #25 from zephyrproject-rtos/main

Merge zephyrproject-rtos:main into wenyongh:main

wenyongh committed:
modules: samples: Add wasm-micro-runtime module and samples

Add wasm-micro-runtime as zephyr module and add two samples:
a hello world sample to illustrate how to run WASM application with
wasm-micro-runtime (WAMR) runtime engine, and an app manager
sample to illustrate how to start WAMR app manager in device
side which runs on Zephyr OS and then install/uninstall/query
WASM apps from host side.

Signed-off-by: Wenyong Huang <wenyong.huang@intel.com>

wenyongh commented on samples/wasm_micro_runtime/hello_world/CMakeLists.txt:
```diff
@@ -0,0 +1,49 @@
+
+if (NOT BOARD)
+  set (BOARD qemu_x86_nommu)
+endif ()
```
Done, removed.

wenyongh commented on samples/wasm_micro_runtime/hello_world/CMakeLists.txt:
```diff
@@ -0,0 +1,49 @@
+  set (BOARD qemu_x86_nommu)
+endif ()
+
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Done.

wenyongh commented on samples/wasm_micro_runtime/hello_world/CMakeLists.txt:
```diff
@@ -0,0 +1,49 @@
+# if we want to support arm, thumb, mips or xtensa
+if (NOT DEFINED WAMR_BUILD_TARGET)
+  set (WAMR_BUILD_TARGET "X86_32")
+endif ()
```
Done, we added CONFIG_WAMR_BUILD_TARGET to allow user to specify the build target, and it isn't set, the build scripts will configure it automatically according to the CPU architecture info. Normally developer doesn't need to set it.

wenyongh commented on samples/wasm_micro_runtime/hello_world/README.rst:
```diff
@@ -0,0 +1,114 @@
+
+    west build -b nucleo_f767zi . -p always -- \
+               -DCONF_FILE=target_configs/prj_nucleo767zi.conf \
+               -DWAMR_BUILD_TARGET=THUMBV7
```
Done, -DWAMR_BUILD_TARGET=xxx was removed.

wenyongh commented on samples/wasm_micro_runtime/hello_world/CMakeLists.txt:
```diff
@@ -0,0 +1,49 @@
+if (NOT DEFINED WAMR_BUILD_LIBC_WASI)
+  # Disable libc wasi support by default
+  set (WAMR_BUILD_LIBC_WASI 0)
+endif ()
```
Done, the related configs were moved to wasm-micro-runtime module's Kconfig.

wenyongh commented:
> @wenyongh thanks for this contribution. A good start, but needs some additional effort.
> 
> I agree with @carlescufi that this should be more generic, and for a start I have placed a few comments for you to take a look at.

@tejlmand @carlescufi Many thanks for your suggestions, we have updated the PR to make WAMR as a zephyr module and updated the related samples' build scripts and documents, could you please help review the patch again? And could you please fork [WAMR](https://github.com/bytecodealliance/wasm-micro-runtime) into [zephyrproject-rtos](https://github.com/zephyrproject-rtos) so that we can update zephyr/west.yml? Thanks.

wenyongh commented:
> @wenyongh one thing I am missing here, where is the actual code being pulled from? We should likely be adding an entry to `west.yml`? Or is one supposed to download the runtime separately?

Yes, it supposes that the WAMR source code is forked into https://github.com/zephyrproject-rtos and west will clone it under zephyrproject/modules/lib. I have updated zephyr/west.yml to add wasm-micro-runtime into project list. The patch should work after [WAMR](https://github.com/bytecodealliance/wasm-micro-runtime) project is cloned.

wenyongh committed:
samples: Change wasm-micro-runtime samples' name

Change sample hello_world to wasm_hello_world, and
change sample simpe to wasm_app_manager.

Signed-off-by: Wenyong Huang <wenyong.huang@intel.com>

wenyongh committed:
west.yml: Add wasm-micro-runtime into projects

Add wasm-micro-runtime into west.yml projects.

Signed-off-by: Wenyong Huang <wenyong.huang@intel.com>

tejlmand reviewed (changes requested):
First round of comments

wenyongh committed:
samples: Enable build host_tool and wasm apps with cmake

Enable build host_tool and wasm apps with cmake, and
update west.yml.

Signed-off-by: Wenyong Huang <wenyong.huang@intel.com>

wenyongh committed:
samples: Refactor CMakeLists.txt of app manager sample

Build wamr-sdk and host-tool by ExternalPorject_Add,
detect wasi-sdk with $ENV{WASI_SDK_PATH} and find_program,
replace some shell commands with cmake commands.

Signed-off-by: Wenyong Huang <wenyong.huang@intel.com>

wenyongh committed:
samples: add include path and update west.yml

Add libc-builtin-sysroot/include to include path when
building wasm apps.
Update the commit ID of wasm-micro-runtime in west.yml.

Signed-off-by: Wenyong Huang <wenyong.huang@intel.com>

wenyongh committed:
samples: Sync up with upstream and update west.yml

Sync up with upstream to get the latest code and update the commit ID
of wasm-micro-runtime in west.yml.

Signed-off-by: Wenyong Huang <wenyong.huang@intel.com>

ivan2angelov commented:
I am very interested in studying WASMR as part of Zephyr. This PR looks almost complete. Are there any plans to merge it?
bot closed it due to lack of activity, @wenyongh shall this PR be reopen?

PureEngineering commented:
I don't understand why this wasn't merged in? can someone revisit this.

xwang98 commented:
This is the Zephyr team's call for accepting this PR or not.  @ivan2angelov @PureEngineering

ivan2angelov commented:
@xwang98 do you have some contact to Zephyr's team? Is it possible to raise an issue to adding WAMR to Zephyr?

soburi commented:
@wenyongh @carlescufi @tejlmand 

I'm interested in this PR.
At a glance, all pointing outs are solved.
And I verified this PR can work on the newest version of Zephyr with few modifications.

So, what reason to reject this PR at that time?
I will help to solve if there are problems.
